src/common/mem.c

```diff
@@ -1696,182 +1703,454 @@ struct umem_page_info {
 	uint64_t pi_last_checkpoint;
 	/** Highest transaction ID of writes to the page */
 	uint64_t pi_last_inflight;
-	/** link chain on global dirty list, LRU list, or free info list */
-	d_list_t pi_link;
+	/** link to global LRU lists, or global free page list, or global pinned list */
+	d_list_t pi_lru_link;
+	/** link to global dirty page list, or wait commit list, or temporary list for flushing */
+	d_list_t pi_dirty_link;
+	/** link to global flushing page list */
+	d_list_t pi_flush_link;
+	/** Waitqueue for page loading/flushing */
+	void	*pi_io_wq;
+	/** Waitqueue for page committing */
+	void	*pi_commit_wq;
 	/** page memory address */
 	uint8_t *pi_addr;
 	/** Information about in-flight checkpoint */
 	void    *pi_chkpt_data;
-	/** bitmap for each dirty 16K unit */
-	uint64_t pi_bmap[UMEM_CACHE_BMAP_SZ];
+	/** bitmap for each dirty 4K unit */
+	uint64_t *pi_bmap;
 };
 
-int
-umem_cache_alloc(struct umem_store *store, uint64_t max_mapped)
+/* Convert MD-blob offset to MD page */
+static inline struct umem_page *
+cache_off2page(struct umem_cache *cache, umem_off_t offset)
 {
-	struct umem_cache *cache;
-	struct umem_page_info *pinfo;
-	uint64_t           num_pages;
-	int                rc = 0;
-	int                idx;
+	uint32_t idx = offset >> cache->ca_page_shift;
```

Comment:
The allocator will use the first 4k bytes for heap header and this will not be part of the memory bucket. The actual buckets starts at offset 4096. To handle this can we add one more field cache->ca_base_offset which indicates the offset where zone0 starts and change the above id calculation as (offset - cache->ca_base_offset) >> cache->ca_page_shift. Similar change is required at other places where the offset is used for translation.

Reply:
You mean the this 4k header is part of bucket 0 but would never be accessed by VOS (The offset returned to VOS starts from 4k)? Then I don't see why umem cache needs this tweak for page ID calculation, I think it just need to ensure never load/flush the first 4k of page 0, right?

If bucket 0 starts from 4k (the 4k header isn't part of bucket 0), then allocator should just use the start of bucket 0 as exported LBA 0, did I miss anything?